src/main/java/com/commercetools/project/sync/product/ProductSyncer.java

```diff
@@ -1,72 +1,53 @@
 
 import com.commercetools.project.sync.Syncer;
 import com.commercetools.project.sync.model.ProductSyncCustomRequest;
-import com.commercetools.project.sync.product.service.ProductReferenceTransformService;
-import com.commercetools.project.sync.product.service.impl.ProductReferenceTransformServiceImpl;
 import com.commercetools.project.sync.service.CustomObjectService;
 import com.commercetools.project.sync.service.impl.CustomObjectServiceImpl;
 import com.commercetools.sync.commons.exceptions.SyncException;
+import com.commercetools.sync.commons.utils.CaffeineReferenceIdToKeyCacheImpl;
 import com.commercetools.sync.commons.utils.QuadConsumer;
+import com.commercetools.sync.commons.utils.ReferenceIdToKeyCache;
 import com.commercetools.sync.commons.utils.TriConsumer;
 import com.commercetools.sync.products.ProductSync;
 import com.commercetools.sync.products.ProductSyncOptions;
 import com.commercetools.sync.products.ProductSyncOptionsBuilder;
 import com.commercetools.sync.products.helpers.ProductSyncStatistics;
-import com.fasterxml.jackson.databind.JsonNode;
-import com.fasterxml.jackson.databind.node.ObjectNode;
-import io.sphere.sdk.categories.Category;
 import io.sphere.sdk.client.SphereClient;
 import io.sphere.sdk.commands.UpdateAction;
-import io.sphere.sdk.customobjects.CustomObject;
-import io.sphere.sdk.products.AttributeContainer;
+import io.sphere.sdk.models.WithKey;
 import io.sphere.sdk.products.Product;
 import io.sphere.sdk.products.ProductDraft;
-import io.sphere.sdk.products.ProductVariant;
-import io.sphere.sdk.products.attributes.Attribute;
-import io.sphere.sdk.products.commands.updateactions.Publish;
-import io.sphere.sdk.products.commands.updateactions.Unpublish;
-import io.sphere.sdk.products.queries.ProductQuery;
-import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.products.ProductProjection;
+import io.sphere.sdk.products.queries.ProductProjectionQuery;
 import io.sphere.sdk.queries.QueryPredicate;
 import java.time.Clock;
-import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
 import java.util.Optional;
-import java.util.Set;
 import java.util.concurrent.CompletionException;
 import java.util.concurrent.CompletionStage;
-import java.util.stream.Collectors;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public final class ProductSyncer
     extends Syncer<
+        ProductProjection,
         Product,
         ProductDraft,
         ProductSyncStatistics,
         ProductSyncOptions,
-        ProductQuery,
+        ProductProjectionQuery,
         ProductSync> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ProductSyncer.class);
-  private static final String REFERENCE_TYPE_ID_FIELD = "typeId";
-  private static final String REFERENCE_ID_FIELD = "id";
-  private static final String WITH_IRRESOLVABLE_REFS_ERROR_MSG =
```

Comment:
In which part of the project-sync and java-sync this case is handled?

Reply:
I think you didn't read the description of this PR. It is good if you have some suggestions on this? It has to be handled first in Java-Sync.

Reply:
I think it does not matter if I read it or not, in this case, I need to provide a suggestion for that, let me check 😄